### PR TITLE
[MTDSA-2104] update reads/writes for TaxCalc

### PR DESCRIPTION
**Pull request by MarkAKelly**

### CODE REVIEW

File: test/v2/models/taxcalculation/TaxCalculationSpec.scala [hunk 1]
```diff
@@ -17,102 +17,104 @@
...
+        expectedError = JSARRAY_FORMAT_EXCEPTION
+      )
+      testPropertyType[TaxCalculation](TestData.v3_2DesTaxCalcJson)(
+        path = "/calcOutput/calcResult/msgCount",
+        replacement = "test".toJson,
+        expectedError = JSNUMBER_FORMAT_EXCEPTION
+      )
     }
 
-    val validJsonPayloadsForVersion3_2 = Map(
```
Comment (macgrewal) [inline test/v2/models/taxcalculation/TaxCalculationSpec.scala]: If tests fixures are no longer used please remove them

File: test/v2/models/taxcalculation/TaxCalculationSpec.scala [hunk 2]
```diff
@@ -17,102 +17,104 @@
...
+            | "intentToCrystallise": false,
+            | "crystallised": false,
+            | "validationMessageCount": 5
+            |}
+          """.stripMargin)
 
-      eoyEstimateJson.as[EoyEstimate](EoyEstimate.reads) shouldBe expectedEoyEstimate
+        TaxCalculation.writes.writes(bvrErrorTaxCalc) shouldBe clientJson
+      }
+      "all fields are present" in {
```
Comment (macgrewal) [inline test/v2/models/taxcalculation/TaxCalculationSpec.scala]: Add writes tests for each instance of an optional val being `None`

File: test/v2/models/taxcalculation/TaxCalculationSpec.scala [hunk 3]
```diff
@@ -17,102 +17,104 @@
...
-    "JSON representing UK property with FHL with periodic and annual updates" -> TestData.ukPropertyFhlPeriodicAndAnnualsJson,
-    "JSON representing Scottish self employment with multiple income sources" -> TestData.selfEmploymentScottishMultipleIncomeSourcesJson
-    )
+    "parse the correct fields when at least one 'bvrError' is returned" in {
+      TaxCalculation.reads.reads(TestData.bvrErrorJson).get shouldEqual bvrErrorTaxCalc
+    }
 
-    validJsonPayloadsForVersion3_2.foreach{
-      case (description, payload) => s"be able to parse $description without falling over" in { payload.as[TaxCalculation] }
+    "parse the correct fields when all fields are present" in {
```
Comment (macgrewal) [inline test/v2/models/taxcalculation/TaxCalculationSpec.scala]: Add reads tests for each scenario where an optional property is missing

File: test/v2/models/utils/JsonErrorValidators.scala [hunk 4]
```diff
@@ -185,6 +185,27 @@ trait JsonErrorValidators {
 
   }
 
+  def amendJson[T](json: JsValue)(path: String, replacement: JsValue)(implicit rds: Reads[T]): T = {
```
Comment (macgrewal) [inline test/v2/models/utils/JsonErrorValidators.scala]: the `amendJson` function is actually used to remove a field from a `JsObject` please update it's name and then the implementation becomes much simpler like the below:
```
  def removeJsonProperty(json: JsValue, pathToProperty: String): JsObject = {
    val jsPath = pathToProperty.split("/").filterNot(_ == "").foldLeft(JsPath())(_ \ _)

    jsPath.prune(json).fold(
      invalid = errs => fail(s"an error occurred when reading $pathToProperty : $errs"),
      valid = x => x
    )
  }
```

File: test/v2/models/taxcalculation/TaxCalculationSpec.scala [hunk 5]
```diff
@@ -16,103 +16,238 @@
...
+
+      "the 'nationalRegime' field is not received" in {
+        val model = amendJson[TaxCalculation](json)(path = "calcOutput/calcResult/nationalRegime", replacement = JsNull)
+        model.nationalRegime shouldBe None
+      }
 
-    validJsonPayloadsForVersion3_2.foreach{
-      case (description, payload) => s"be able to parse $description without falling over" in { payload.as[TaxCalculation] }
+      "the 'totalBeforeTaxDeducted' field is not received" in {
+        val model = amendJson[TaxCalculation](json)(path = "calcOutput/calcResult/totalBeforeTaxDeducted", replacement = JsNull)
```
Comment (macgrewal) [inline test/v2/models/taxcalculation/TaxCalculationSpec.scala]: These tests says *"the 'X' field is not received"*, in truth the `amendJson` function doesn't remove to field it sets the value to be `null`. A comment exists against that function to reflect this point and once it is resolved please re-write test implementations to use newly name function in `JsonErrorValidators` e.g.
```
"the 'totalBeforeTaxDeducted' field is not received" in {
  val updatedJson = removeJsonProperty(json, "calcOutput/calcResult/totalBeforeTaxDeducted")
  val model = updatedJson.as[TaxCalculation]
  model.totalBeforeTaxDeducted shouldBe None
}
```
